### ISSUE TITLE
Query returns `undefined` when same cache object is in multiple query responses

### ISSUE THREAD

**Issue by irace** [apollographql/react-apollo#2678]
I have two queries in components that are on screen at the same time – one (`me`) returns the current user, and another (`group`) returns all members of a group. Once cached, the cached results of each query look as follows:

```
ROOT_QUERY
  me: User
    User:bryan-irace
  group({"id": 1}): Group
    Group:some-group-name
      members: GroupMembers
        results: [User]
          0: User:bryan-irace
          1: User:kevin-grant
```

I am seeing some very strange behavior when the `User` model returned by the `me` query is also included in the `members.results` array returned by the `group` query. Specifically, the `me` query’s callback is invoked multiple times:

1. `data.me` returned `undefined` (network request has not occurred and cache has not been populated yet)
2. `data.me` returns the correct model (model is now stored in the cache)
3. `data.me` returns `undefined again`, despite the model being present on disk

This third invocation is problematic – I don’t know why it is being triggered to begin with, but if it’s going to be triggered, `data.me` should certainly be populated.

I can fix this problem by either:

1. Removing the `groups` query altogether
2. Modifying the data such that the user returned by the `me` query is no longer a group member

Obviously neither of these are suitable workarounds, but in both cases, the third `me` query callback invocation does not occur.

Please advise 🙏

**Comment by irace** [comment]
If it helps, here are the two different variations on the same model, the first from the `me` query and the second from the `group` query:

```json
"admin": true,
"firstName": "Bryan",
"lastName": "Irace",
"bio": "Push code too late on a Friday – what is wrong with me",
"professions": [{
  "id": 2,
  "name": "Software Developer",
  "__typename": "Profession"
}, {
  "id": 8,
  "name": "Polyglot Engineer42",
  "__typename": "Profession"
}],
"phoneNumber": "+16314959276",
"city": "Chicago, NY",
"email": "bryan@foo.bar",
"website": "https://irace.me",
"linkedinUrl": null,
"facebookUrl": null,
"instagramUrl": null,
"twitterUrl": null,
"googleUrl": null,
"yelpUrl": null,
"foursquareUrl": null,
"slug": "bryan-irace",
"tags": [],
"avatar": {
  "large": "https://prefer3-staging.herokuapp.com/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWjg9IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--82ddb463198c7998545b01ee4a438bf16e00ed0b/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9MY21WemFYcGxTU0lNTXpBd2VETXdNQVk2QmtWVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--2726c128c71a8b3a495028e9c2cc6567207d11ef/blob",
  "thumbnail": "https://prefer3-staging.herokuapp.com/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWjg9IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--82ddb463198c7998545b01ee4a438bf16e00ed0b/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9MY21WemFYcGxTU0lLT1RaNE9UWUdPZ1pGVkE9PSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--7406d2a64a55bfbdcb289d4867d8553dd3235704/blob",
  "__typename": "UserAvatar"
}
```

```json
"firstName": "Bryan",
"lastName": "Irace",
"professions": [{
  "name": "Software Developer",
  "__typename": "Profession"
}, {
  "name": "Polyglot Engineer42",
  "__typename": "Profession"
}],
"slug": "bryan-irace",
"avatar": {
  "large": "https://prefer3-staging.herokuapp.com/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWjg9IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--82ddb463198c7998545b01ee4a438bf16e00ed0b/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9MY21WemFYcGxTU0lNTXpBd2VETXdNQVk2QmtWVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--2726c128c71a8b3a495028e9c2cc6567207d11ef/blob",
  "thumbnail": "https://prefer3-staging.herokuapp.com/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWjg9IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--82ddb463198c7998545b01ee4a438bf16e00ed0b/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9MY21WemFYcGxTU0lLT1RaNE9UWUdPZ1pGVkE9PSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--7406d2a64a55bfbdcb289d4867d8553dd3235704/blob",
  "__typename": "UserAvatar"
},
"tags": [],
"__typename": "User"
}
```

**Comment by sbrichardson** [comment]
Is there an id field that you left off? Not having an id field will cause a lot of issues.

**Comment by irace** [comment]
@sbrichardson Thanks for getting back to me. In this case, the `slug` field is the unique identifier used for caching purposes.

**Comment by irace** [comment]
This is a pretty easy mistake to make though, and the fact that this just returns `undefined` when the query’s data is visible in the cache via the Apollo inspector makes it really hard to figure out what's going on. No errors or anything.

**Comment by sbrichardson** [comment]
Yes I've ran into that issue, was going to mention, typically it manifests if you remove an id field from a selection set accidentally, such as in a table UI, editing table columns to build a gql query dynamically.

**Comment by irace** [comment]
@sbrichardson That documentation that you linked to doesn’t really give any indication that a missing cache key on a child object would prevent the parent object from being properly normalized. It seems like this should theoretically be something that Apollo can handle, or at the very least provide a more helpful error message?

**Comment by developdeez** [comment]
Been stuck on this for hours. What's the fix? 
It happens almost randomly it seems.

**Comment by Elijen** [comment]
@developdeez I've fixed this by making sure `id` field is always selected for all nodes (might be nice to have this done automatically if omitting a field breaks the app). I am also now re-using the same query for components that request the same object. This kind of defeats the _"only request what you really need"_ advantage of GraphQL, but is not a big deal for me.

**Comment by hwillson** [comment]
See the `react-apollo` 2.5.6 [release notes](https://github.com/apollographql/react-apollo/blob/master/Changelog.md#256-2019-05-22) for details around the `returnPartialData` prop. It will help here. Thanks!

**Comment by vfonic** [comment]
@hwillson I don't think that `returnPartialData` resolves the issue.

`returnPartialData` provides data from the cache immediately. Here the issue is different: after the query data is returned from the backend, instead of it being available in `data` property, the `data` property is `undefined`.

There's no error message and this is really frustrating to debug. I've checked already and the backend correctly returns the data.

As @Elijen said (emphasis mine):
> by making sure `id` field is always selected for all nodes (might be nice to have this done automatically **if omitting a field breaks the app**)

I've added `id` field everywhere, but the data still doesn't show. I guess I'll have to use the same query everywhere, basically using GraphQL the same way I used to use REST API. :/ Kind of defeats the purpose, don't you think?
